### PR TITLE
Modify the overall architecture and add low-frequency storage support

### PR DESCRIPTION
**Pull request by fireworkmarks**

### CODE REVIEW

File: billing/postbilling.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+	if Conf.EnableUsageCache {
+		wg.Add(1)
+		go task.cacheUsageToRedis(wg)
+	}
+
+	// Constuct BillingData
+	task.BillingData.RegionId = Conf.RegionId
+	hd, _ := time.ParseDuration("-1h")
+	task.BillingData.BeginTime = time.Now().Add(hd).Format(spark.TimeLayoutStr) + ":00:00"
+	task.BillingData.EndTime = time.Now().Format(spark.TimeLayoutStr) + ":00:00"
```
Comment (cannium) [inline billing/postbilling.go:79]: Format with `Format`, I don't think the string concatenation is necessary.

File: billing/collectmessage.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		}
+	} else {
+		object.Size = bodyBytes
+		object.CountSize = getCountSize(object.Size)
+		err := db.DbClient.InsertBilling(*object)
+		if err != nil {
+			Logger.Println("[MESSAGE ERROR] Del Put Object Error with database insert, err:", err, "Uuid is :", msg.Uuid)
+			return
+		}
+	}
```
Comment (cannium) [inline billing/collectmessage.go]: Just use `bodyBytes` as `object.Size`...

File: billing/collectmessage.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	if err != nil {
+		Logger.Println("[MESSAGE ERROR] Del Delete Object Error with database update, err:", err, "Uuid is :", msg.Uuid)
+		return
+	}
+	Logger.Println("[MESSAGE] Delete object successful, Uuid is:", msg.Uuid)
+}
+
+// Round up for the smallest unit
+func getCountSize(objectSize int64) (countSize int64) {
+	countSize = objectSize - (objectSize % MinSizeForStandardIa) + MinSizeForStandardIa
```
Comment (cannium) [inline billing/collectmessage.go]: You mean `countSize = (objectSize + MinSizeForStandardIa - 1) / MinSizeForStandardIa * MinSizeForStandardIa`?

File: billing/collectmessage.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	err := db.DbClient.UpdateBillingFlag(*object)
+	if err != nil {
+		Logger.Println("[MESSAGE ERROR] Del Delete Object Error with database update, err:", err, "Uuid is :", msg.Uuid)
+		return
+	}
+	Logger.Println("[MESSAGE] Delete object successful, Uuid is:", msg.Uuid)
+}
+
+// Round up for the smallest unit
+func getCountSize(objectSize int64) (countSize int64) {
```
Comment (cannium) [inline billing/collectmessage.go]: Would `getBillingSize` more understandable?

File: billing/postbilling.go (hunk 1 above)
Comment (fireworkmarks) [reply billing/postbilling.go:79]: This is used to mark the time period

File: spark/fetch_usage.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+path=$1
+SPARKHOME=$2
+
+exec $SPARKHOME/bin/spark-shell <<!EOF
+spark.sql("use yig")
+val df1 = sql("select ownerid,storageclass,sum(size) as usage from objects where storageclass=0 group by ownerid,storageclass")
+val df2 = sql("select m.ownerid,m.storageclass, sum(mp.size) from multipartpart mp left join multiparts m on m.bucketname = mp.bucketname and m.uploadtime=mp.uploadtime group by m.ownerid,m.storageclass").where("storageclass=0")
+val df3 = df1.unionAll(df2).distinct().groupBy("ownerid","storageclass").sum("usage")
+val df4 = sql("select ownerid,storageclass,sum((size + 65536 - 1) / 65536 * 65536) as usage from objects where storageclass!=0 group by ownerid,storageclass")
```
Comment (cxt90730) [inline spark/fetch_usage.sh]: =3

File: billing/collectmessage.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	. "github.com/journeymidnight/yig-billing/helper"
+	"github.com/journeymidnight/yig-billing/messagebus"
+	"github.com/journeymidnight/yig-billing/redis"
+	"strconv"
+)
+
+const (
+	DEADLINE             = 2592000
+	UNDERLINE            = ":"
```
Comment (cxt90730) [inline billing/collectmessage.go]: Seperator

File: billing/postbilling.go [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+		Usage:    usageCount,
+	}
+	if _, ok := t.PidCache[pid]; ok {
+		t.PidCache[pid] = append(t.PidCache[pid], usage)
+	} else {
+		t.PidCache[pid] = []BillingUsage{usage}
+	}
+}
+
+func (t *Task) ConstructUsageOtherStorageClassBeenDeletedData() {
```
Comment (cxt90730) [inline billing/postbilling.go]: ConstructIADeletedUsage

**Review by cxt90730 — APPROVED** [review]

File: billing/collectmessage.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	. "github.com/journeymidnight/yig-billing/helper"
+	"github.com/journeymidnight/yig-billing/messagebus"
+	"github.com/journeymidnight/yig-billing/redis"
+	"strconv"
+)
+
+const (
+	DEADLINE             = 2592000
+	SEPERATOR            = ":"
```
Comment (cannium) [inline billing/collectmessage.go]: “SEPARATOR” 😂

File: redis/redis.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func (r *Redis) SetToRedis(message MessageForRedis) {
+	conn := r.pool.Get()
+	_, err := conn.Do("SET", message.Key, message.Value)
+	if err != nil {
+		Logger.Println("[WARNING] Redis set error:", err)
+		return
+	}
+	defer conn.Close()
```
Comment (cannium) [inline redis/redis.go]: Put `defer conn.Close()` right after `conn := r.pool.Get()`

File: redis/redis.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func (r *Redis) SetToRedisWithExpire(message MessageForRedis, expire int, uuid string) {
+	conn := r.pool.Get()
+	_, err := conn.Do("SETEX", message.Key, expire, message.Value)
+	if err != nil {
+		Logger.Println("[MESSAGE ERROR] Redis set error:", err, "uuid is:", uuid)
+		return
+	}
+	defer conn.Close()
```
Comment (cannium) [inline redis/redis.go]: same here

File: redis/redis.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+			iter, _ = redis.Int(arr[0], nil)
+			keys, _ = redis.Strings(arr[1], nil)
+		}
+		allKeys = append(allKeys, keys...)
+		// check if we need to stop...
+		if iter == 0 {
+			break
+		}
+	}
+	defer conn.Close()
```
Comment (cannium) [inline redis/redis.go]: and here

File: redis/redis.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	return
+}
+
+func (r *Redis) GetFromRedis(key string) (result string) {
+	conn := r.pool.Get()
```
Comment (cannium) [inline redis/redis.go:92]: and here...

File: billing/collectmessage.go (hunk 8 above)
Comment (fireworkmarks) [reply billing/collectmessage.go]: Ok， I made a mistake！

File: redis/redis.go (hunk 9 above)
Comment (fireworkmarks) [reply redis/redis.go]: roger

**Review by cannium — APPROVED** [review]

**Review by cxt90730 — APPROVED** [review]

**Pull request merged**